lib/bibliothecary/analyser.rb

```diff
@@ -42,12 +42,40 @@ def map_dependencies(hash, key, type)
       end
 
       def analyse(folder_path, file_list)
-        file_list.map do |path|
+        analyses = file_list.map do |path|
           filename = path.gsub(folder_path, '').gsub(/^\//, '')
           next unless match?(filename)
           contents = File.open(path).read
           analyse_contents(filename, contents)
         end.compact
+
+        analyses.each do |analysis|
+          analysis[:related_paths] = []
+        end
+
+        # associate manifests and lockfiles in the same directory;
+        # note that right now we're in the context of a single
+        # package manager, so manifest and lockfile in the
+        # same directory is considered proof that they are
+        # matched.
+        by_dirname = {}
+        analyses.each do |analysis|
+          dirname = File.dirname(analysis[:path])
+          by_dirname[dirname] ||= []
+          by_dirname[dirname].push(analysis)
+        end
+
+        # set related_paths so manifests point to lockfiles and vice versa
+        by_dirname.each do |_, all_for_dirname|
+          manifest = all_for_dirname.find { |analysis| analysis[:kind] == "manifest" }
```

Comment:
I wonder if it's worth having two hashes to start with... one that's manifests_by_dirname and one that's lockfiles_by_dirname. Seems like it'd save some iteration on really deep trees and also might be a little cleaner.